cluster-autoscaler/core/scaledown/actuation/priority_evictor.go

```diff
@@ -0,0 +1,238 @@
+// All non-mirror, non-DS pods on the node are evicted. Mirror pods are not evicted.
+// DaemonSet pods are evicted if DaemonSetEvictionForOccupiedNodes is enabled, or if they have the EnableDsEvictionKey annotation.
+func (e PriorityEvictor) DrainNode(ctx *context.AutoscalingContext, nodeInfo *framework.NodeInfo) (map[string]status.PodEvictionResult, error) {
+	pods := podsToEvictCombined(ctx, nodeInfo)
+	return e.DrainNodeWithPodsBasedOnPodPriority(ctx, nodeInfo.Node(), pods)
+}
+
+// EvictDaemonSetPods creates eviction objects for all DaemonSet pods on the node.
+func (e PriorityEvictor) EvictDaemonSetPods(ctx *context.AutoscalingContext, nodeInfo *framework.NodeInfo, timeNow time.Time) error {
+	return evictDaemonSetPods(ctx, nodeInfo, timeNow, e.deleteOptions, e.DsEvictionEmptyNodeTimeout, e.DsEvictionRetryTime, e.evictionRegister)
+}
+
+// DrainNodeWithPodsBasedOnPodPriority performs drain logic on the node based on pod priorities. Marks the node as unschedulable and
+// later removes all pods, giving them up to MaxGracefulTerminationTime to finish. The list of pods to evict has to be provided.
+func (e PriorityEvictor) DrainNodeWithPodsBasedOnPodPriority(ctx *context.AutoscalingContext, node *apiv1.Node, pods []*apiv1.Pod) (map[string]status.PodEvictionResult, error) {
+	evictionResults := make(map[string]status.PodEvictionResult)
+	retryUntil := time.Now().Add(ctx.MaxPodEvictionTime)
+
+	groups := groupByPriority(e.shutdownGracePeriodByPodPriority, pods)
+	for _, group := range groups {
+		for _, pod := range group.Pods {
+			evictionResults[pod.Name] = status.PodEvictionResult{Pod: pod, TimedOut: true,
+				Err: errors.NewAutoscalerError(errors.UnexpectedScaleDownStateError, "Eviction did not attempted for the pod %s because some of the previous evictions failed", pod.Name)}
+		}
+	}
+
+	for _, group := range groups {
+		// If there are no pods in a particular range,
+		// then do not wait for pods in that priority range.
+		if len(group.Pods) == 0 {
+			continue
+		}
+		confirmations := make(chan status.PodEvictionResult, len(group.Pods))
+		for _, pod := range group.Pods {
+			evictionResults[pod.Name] = status.PodEvictionResult{Pod: pod, TimedOut: true, Err: nil}
+			go func(pod *v1.Pod, group podShutdownGroup) {
+				gracePeriodOverride := group.ShutdownGracePeriodSeconds
+
+				// If the pod's spec specifies a termination gracePeriod which is less than the gracePeriodOverride calculated, use the pod spec termination gracePeriod.
+				if pod.Spec.TerminationGracePeriodSeconds != nil && *pod.Spec.TerminationGracePeriodSeconds <= gracePeriodOverride {
+					gracePeriodOverride = *pod.Spec.TerminationGracePeriodSeconds
+				}
+				confirmations <- evictPod(ctx, pod, retryUntil, e.EvictionRetryTime, e.evictionRegister, gracePeriodOverride)
+			}(pod, group)
+		}
+
+		podsEvictionCounter := 0
+		for i := 0; i < len(group.Pods); i++ {
```

Comment:
`evictPod` doesn't wait for the pod to disappear before returning - it returns immediately after it successfully initiates the eviction. If this is supposed to imitate the kubelet node shutdown behavior, we should wait until all pods from a given group disappear before moving on to evicting the next group - otherwise the pods are still effectively evicted at the same time, just in batches.

Btw, this also seems a bit easier to implement if we merge the two Evictor implementations.

Reply:
Done